bazelisk.go

```diff
@@ -539,12 +539,29 @@ func insertArgs(baseArgs []string, newArgs []string) []string {
 	return result
 }
 
+func clean(bazelPath string) {
+	bazeliskNoClean := os.Getenv("BAZELISK_NO_CLEAN")
```

Comment:
We might want to flip this instead and have users opt in since it could be a big time difference

Reply:
Agree.

Reply:
The question is: While running migrate without cleaning is obviously faster, does it give any usable results? In theory it should, in practice I’m not at all convinced. 

Incompatible flag flips that don’t change action cache keys will not cause any actions to rerun, even though there might be differences, e.g. in how execution strategy selection works, how execution strategies do their stuff, memory consumption...

What’s our history of detecting and missing breakages with the current behavior?

Reply:
The current behavior actually gives us pretty accurate result in the [incompatible change pipeline](https://buildkite.com/bazel/bazelisk-plus-incompatible-flags/builds/162). I had the same concern when I switched incompatible flags pipeline implementation from running every flag with every project in an individual job to running Bazelisk for all flags in one job. But ever since then, I saw only two times Bazelisk gave wrong result because of not doing bazel clean. The later one actually helped us discover a [bug](https://github.com/bazelbuild/bazel/issues/8693).

Reply:
If we get incorrect results without clean, that's a bug in Bazel. I'd prefer that we file bugs in that case.

Reply:
> If we get incorrect results without clean, that's a bug in Bazel. I'd prefer that we file bugs in that case.

Yes, I agree with that. I believe that all changes to execution strategies will not be adequately tested without doing a "clean". I don't know about other areas, maybe things look better there.

But if you think it's fine to make this opt-in (= tell the user they have to make the decision), then under which circumstances should users opt-in and under which would it be fine to not do so?

Unless we can clearly answer this, I think it's wrong to make this configurable - either doing it without "clean" is OK, then we do it without clean, or it's not OK, then we always do it with "clean" (even though it's slower). How should a user make the correct decision here if not even we can?

Reply:
I've filed https://github.com/bazelbuild/bazel/issues/8787 to track the specific issue of incompatible_enable_cc_toolchain_resolution.
I've sent an email (https://groups.google.com/forum/#!topic/bazel-dev/UJubXrYxHVM) to discuss the root cause.

Reply:
+1 to what Laurent said, let's fix the root cause and keep Bazelisk assuming Bazel is correct.

Reply:
Based on this conversation should we keep the default the same and still add this, or not add this at all until we have a better understanding of this specific issue?

Reply:
Thanks for the productive discussion everyone :)

@keith Let’s add the flipped version!

Reply:
Flipped!